Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp

```diff
@@ -0,0 +1,261 @@
+  }
+  else
+  {
+    // only horizontal borders
+    auto diff = dstSize.width - w2;
+    if (diff == 1)
+    {
+      left = 1;
+      right = 0;
+    }
+    else
+    {
+      left = diff / 2;
+      right = left;
+    }
+    cv::resize(input, output, cv::Size(w2, dstSize.height));
+  }
+
+  cv::copyMakeBorder(output, output, top, down, left, right, cv::BORDER_CONSTANT, bgcolor);
+
+  return output;
+}
+}  // namespace
+
+namespace InputCommon
+{
+DynamicInputTextureConfiguration::DynamicInputTextureConfiguration(const std::string& json_file)
+{
+  std::ifstream json_stream(json_file);
+  if (!json_stream.is_open())
+  {
+    m_valid = false;
+  }
+  std::string json_str((std::istreambuf_iterator<char>(json_stream)),
+                       (std::istreambuf_iterator<char>()));
+
+  picojson::value out;
+  auto error = picojson::parse(out, json_str);
+
+  if (!error.empty())
+  {
+    m_valid = false;
+    return;
+  }
+
+  // Required fields
+  picojson::value& colorkey_image = out.get("colorkey_image");
+  picojson::value& result_image = out.get("result_image");
+  picojson::value& device = out.get("device");
+  picojson::value& emulated_controls = out.get("emulated_controls");
+  picojson::value& host_controls = out.get("host_controls");
+
+  if (!colorkey_image.is<std::string>() || !result_image.is<std::string>() ||
+      !device.is<std::string>() || !emulated_controls.is<picojson::object>() ||
+      !host_controls.is<picojson::object>())
+  {
+    m_valid = false;
+    return;
+  }
+
+  m_colorKeyImageName = colorkey_image.to_str();
+  m_hiResTextureName = result_image.to_str();
+  m_devicename = device.to_str();
+
+  SplitPath(json_file, &m_base_path, nullptr, nullptr);
+  if (!File::Exists(m_base_path + m_colorKeyImageName))
+  {
+    m_valid = false;
+    return;
+  }
+
+  const auto& emulated_controls_json = emulated_controls.get<picojson::object>();
+  for (auto& [emulated_control, color] : emulated_controls_json)
+  {
+    const int color_num = std::stoi(color.to_str().substr(1), 0, 16);
+    ColorKey color_key;
+    color_key.r = color_num / 0x10000;
+    color_key.g = (color_num / 0x100) % 0x100;
+    color_key.b = color_num % 0x100;
+
+    m_emulated_input_to_color.insert_or_assign(emulated_control, std::move(color_key));
+  }
+
+  const auto& host_controls_json = host_controls.get<picojson::object>();
+  for (auto& [host_control, image_name] : host_controls_json)
+  {
+    m_host_input_to_input_image.insert_or_assign(host_control, image_name.to_str());
+  }
+
+  // Optional fields
+  picojson::value& style = out.get("style");
+
+  if (style.is<std::string>())
+  {
+    m_style = style.to_str();
+  }
+
+  picojson::value& preserve_aspect_ratio = out.get("preserve_aspect_ratio");
+
+  if (preserve_aspect_ratio.is<bool>())
+  {
+    m_preserve_aspect_ratio = preserve_aspect_ratio.get<bool>();
+  }
+}
+
+void DynamicInputTextureConfiguration::GenerateTexture(IniFile::Section* sec)
```

Comment:
`sec` can be a `const IniFile::Section*`, given it only queries the section contents. (ditto for `GenerateTextures`)